ghcide/src/Development/IDE/Plugin/Completions/Logic.hs

```diff
@@ -560,24 +560,74 @@ getCompletions
     -> ClientCapabilities
     -> CompletionsConfig
     -> HM.HashMap T.Text (HashSet.HashSet IdentInfo)
-    -> IO [Scored CompletionItem]
-getCompletions plId ideOpts CC {allModNamesAsNS, anyQualCompls, unqualCompls, qualCompls, importableModules}
-               maybe_parsed (localBindings, bmapping) prefixInfo caps config moduleExportsMap = do
-  let VFS.PosPrefixInfo { fullLine, prefixModule, prefixText } = prefixInfo
-      enteredQual = if T.null prefixModule then "" else prefixModule <> "."
+    -> [Scored CompletionItem]
+getCompletions
+    plId
+    ideOpts
+    CC {allModNamesAsNS, anyQualCompls, unqualCompls, qualCompls, importableModules}
+    maybe_parsed
+    (localBindings, bmapping)
+    VFS.PosPrefixInfo {fullLine, prefixModule, prefixText, cursorPos}
+    caps
+    config
+    moduleExportsMap
+    -- ------------------------------------------------------------------------
+    -- IMPORT MODULENAME (NAM|)
+    | Just (ImportListContext moduleName) <- maybeContext
+    = moduleImportListCompletions moduleName
+
+    | Just (ImportHidingContext moduleName) <- maybeContext
+    = moduleImportListCompletions moduleName
+
+    -- TODO: Is manual parsing ever needed or is context always present for module?
```

Comment:
cc @pepeiborra . I don't know how common it is to not have a `ParsedModule` at all. Won't we almost certainly get a stale one? That seems like it should be good enough for e.g. the import context!

Reply:
The caller is using `useWithStaleFast GetParsedModule`: We will always get a parsed module.

Reply:
@pepeiborra At least for import completions that seems* to be true, but if I tried limiting this entire function to `Just (parsedModule, positionMapping)` then some tests did fail. :confused: 
<sub>*) I am a bit afraid we just don't test this enough.</sub>

Reply:
@pepeiborra Well CI proved me wrong - HLS has a test that does not pass without manual parsing:
```
haskell-language-server
  completions
    import function completions:        FAIL (3.23s)
      src/Test/Hls/Util.hs:386:
      Completion with label "Alternative" not found in ["ALRTransitionalBit","ALoad" {- many more -}]
      Use -p '/import function completions/' to rerun this test only.
```
Confusingly though in CI it did pick up some `Alternative` that was a struct:
```
 haskell-language-server
  completions
    import function completions: FAIL (2.09s)
      test/functional/Completion.hs:135:
      expected: Just CiFunction
       but got: Just CiStruct
```

Reply:
The only place where there won't be a parsed module is at file open, before a module has been parsed at least once. This test suite (`test/functional/Completion.hs`) is accidentally exercising that use case non-deterministically, because it doesn't wait for progress done. In comparison, the completion tests in the ghcide test suite (`ghcide/test/exe/Main.hs`) carefully wait for progress done.

Reply:
Thanks for the explanation, I will see if I can get the tests to pass with only parsed module completions :+1: